### DISCUSSION TITLE
AsyncSerialClient: Are request retries implemented?

### DISCUSSION THREAD

**Discussion by aq-daniel** [pymodbus-dev/pymodbus#1506] [Help] [answered]
Hello,
I'm using the AsyncSerialClient to communicate via Modbus RTU with some sensors. For example I'm reading current measurement values every 30 seconds. Sometimes I get no response from a single sensor and the AsyncSerialClient throws an asyncio.TimeoutError after the set timeout (1s). This leads to a conenction loss and the automatic reconnection mechanism of pymodbus starts.
I wonder wether the request retries (which i specified when creating the AsyncSerialClient with retries=3 and retry_on_empty=True) are happening, because I can't find anything related to request retries in the pymodbus debug logs.
Are request retries - for instance sending the same request over the serial line again, if there is no answer from the slave within the timeout - implemented in the AsyncSerialClient?

**Comment by janiversen** [comment]
Depends on the version. It is implemented for all async clients on v3.2 but we are currently refactoring the async client make it better

**Reply by aq-daniel** [reply]
Thank you for working on refactoring and improvements of the async client. :) I'm using version 3.2.2

**Comment by janiversen** [comment, marked as answer]
I was wondering if you have misunderstood how retry works. It is a send retry, which means it will try to resend the message until it is successfully sent or retries have been reached, this has nothing to do with what we receive.

I suspect what you want is something like a transaction_retry, the retries until a response is recieved, but that is something we cannot do in the library, because some write_commands might be dangerous to repeat (e.g. a step_up command) and cause potentially dangerous situations. In your application you know if it is safe to repeat the command.

**Reply by aq-daniel** [reply]
How do you check wether a message has been successfully sent before issuing a retry? I couldn't find it in the async client implementation code - using version 3.2.2. Where are the retries happening, which I set with retries=3 and retry_on_empty=True? Could you kindly point me to that location in the code, please? Is it within the transport? Thanks.

Yes, you are right maybe in some situations/ applications it could be unsafe to retry a modbus transaction. So maybe make the transaction retry optional? Like if I don't want it, I set retries=0. Sorry, if I misunderstood the currently implemented send retry.

As far as i know the Modbus Protocol demands a slave response for every message sent except for broadcast messages. So in my opinion a retry should be done, if there is no slave response and it wasn't a broadcast message. At least I want to implement it like this in my application. What is the best way I could achieve this with the async client in your opinion?
What I find a little bit weird is the complete connection loss after a asyncio.TimeoutError is happening (in AsyncSerialClient.async_execute()). This leads to me needing to wait for a reconnect of the serial line before I can send another modbus request to my slaves and eventually get a response. Or let me put it this way: Although the serial connection doesn't have a problem and only a specific slave is not responding within the timeout, I have to wait for a reconnection to happen before I could send another modbus request.

**Reply by janiversen** [reply]
What can I say, have a look at the code, pull requests are welcome.
